### ISSUE TITLE
Envoy WASM filter proxy_http_call crashes Envoy when destination becomes 'no healthy upstream'

### ISSUE THREAD

**Issue by chandlernine** [istio/istio#31194]
**Bug description**

For Envoy WASM filters, both NETWORK_FILTER and HTTP filters, on Istio 1.5.10, 1.6.13, and 1.9.0, proxy_http_call to an internal destination cluster crashes if the destination becomes 'no healthy upstream'.

[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure
[ ] Upgrade

**Expected behavior**

**Steps to reproduce the bug**

- Install an EnvoyFilter (we install both L7 and L4 filters) and instruct it to make a proxy_http_call to an internal cluster using EDS.
- While that call is outstanding, `kubectl delete --force <all pods in the destination cluster>`
- crash! Below I include two crashes, one for 1.9.0 and one for 1.6.13 (with line numbers):

I.6.13 crash with line numbers:

```
 Caught Segmentation fault, suspect faulting address 0x148
 Backtrace (use tools/stack_decode.py to get line numbers):
 Envoy version: ef96e9c2f0b0d2a31e3a9426c4b339db0b33885f/1.14.5-dev/Clean/RELEASE/BoringSSL
 #0: __restore_rt [0x7f7acf0078a0]->[0x2976a56798a0] ??:0
 #1: Envoy::Http::AsyncRequestImpl::onComplete() [0x56042c40ea8f]->[0x2a80a8f] /usr/lib/llvm/bin/../include/c++/v1/memory:2590
 (inlined by) /usr/lib/llvm/bin/../include/c++/v1/memory:2547
 (inlined by) external/envoy/source/common/http/async_client_impl.cc:270
 #2: Envoy::Http::AsyncStreamImpl::encodeData() [0x56042c40e482]->[0x2a80482] external/envoy/source/common/http/async_client_impl.cc:193
 (inlined by) external/envoy/source/common/http/async_client_impl.cc:118
 #3: Envoy::Http::Utility::sendLocalReply() [0x56042c52d55e]->[0x2b9f55e] bazel-out/k8-opt/bin/external/envoy/source/common/buffer/_virtual_includes/buffer_lib/common/buffer/buffer_impl.h:523
 (inlined by) external/envoy/source/common/http/utility.cc:469
 #4: Envoy::Http::AsyncStreamImpl::sendLocalReply() [0x56042c40f735]->[0x2a81735] /usr/lib/llvm/bin/../include/c++/v1/functional:1818
 (inlined by) /usr/lib/llvm/bin/../include/c++/v1/functional:2406
 (inlined by) bazel-out/k8-opt/bin/external/envoy/source/common/http/_virtual_includes/async_client_lib/common/http/async_client_impl.h:352
 #5: Envoy::Router::Filter::sendNoHealthyUpstreamResponse() [0x56042c41774a]->[0x2a8974a] /usr/lib/llvm/bin/../include/c++/v1/functional:1818
 (inlined by) /usr/lib/llvm/bin/../include/c++/v1/functional:2406
 (inlined by) external/envoy/source/common/router/router.cc:664
 #6: Envoy::Router::Filter::decodeHeaders() [0x56042c415c5c]->[0x2a87c5c] external/envoy/source/common/router/router.cc:550
 #7: Envoy::Http::AsyncStreamImpl::sendHeaders() [0x56042c40e807]->[0x2a80807] external/envoy/source/common/http/async_client_impl.cc:186
 (inlined by) external/envoy/source/common/http/async_client_impl.cc:142
 #8: Envoy::Http::AsyncRequestImpl::initialize() [0x56042c40d7a7]->[0x2a7f7a7] /usr/lib/llvm/bin/../include/c++/v1/memory:2562
 (inlined by) external/envoy/source/common/http/async_client_impl.cc:257
 #9: Envoy::Http::AsyncClientImpl::send() [0x56042c40d20b]->[0x2a7f20b] external/envoy/source/common/http/async_client_impl.cc:59
 #10: Envoy::Extensions::Common::Wasm::Context::httpCall() [0x56042b463142]->[0x1ad5142] external/envoy/source/extensions/common/wasm/context.cc:898
 #11: Envoy::Extensions::Common::Wasm::Exports::http_call() [0x56042b47b0d5]->[0x1aed0d5] external/envoy/source/extensions/common/wasm/exports.cc:561
 #12: Envoy::Extensions::Common::Wasm::ConvertFunctionWordToUint32<>::convertFunctionWordToUint32() [0x56042b486f96]->[0x1af8f96] bazel-out/k8-opt/bin/external/envoy/source/extensions/common/wasm/_virtual_includes/wasm_vm_interface/extensions/common/wasm/wasm_vm.h:55
 #13: Envoy::Extensions::Common::Wasm::V8::V8::registerHostFunctionImpl<>()::{lambda()#1}::operator()() [0x56042b4b1ca2]->[0x1b23ca2] external/envoy/source/extensions/common/wasm/v8/v8.cc:599
 #14: Envoy::Extensions::Common::Wasm::V8::V8::registerHostFunctionImpl<>()::{lambda()#1}::__invoke() [0x56042b4b1b47]->[0x1b23b47] external/envoy/source/extensions/common/wasm/v8/v8.cc:592
 #15: wasm::FuncData::v8_callback() [0x56042b4bd43f]->[0x1b2f43f] ??:?
 #16: [0x193de1441167]
 #17: [0x193de144d185]
 #18: [0x193de144cd86]
 #19: [0x193de1449c6e]
 #20: [0x193de1449b78]
 #21: [0x193de1449a7c]
 #22: [0x193de14424b9]
 #23: [0x21c1000848f0]
 #24: v8::internal::Execution::CallWasm() [0x56042b595f13]->[0x1c07f13] ??:?
 #25: wasm::Func::call() [0x56042b4bc7a3]->[0x1b2e7a3] ??:?
 #26: Envoy::Extensions::Common::Wasm::V8::V8::getModuleFunctionImpl<>()::{lambda()#1}::operator()() [0x56042b4a2fb1]->[0x1b14fb1] /usr/lib/llvm/bin/../include/c++/v1/memory:2578
 (inlined by) external/envoy/source/extensions/common/wasm/v8/v8.cc:658
 #27: Envoy::Extensions::Common::Wasm::Context::onDone() [0x56042b468582]->[0x1ada582] external/envoy/source/extensions/common/wasm/context.cc:1647
 #28: Envoy::Http::ConnectionManagerImpl::doDeferredStreamDestroy() [0x56042c47bf6e]->[0x2aedf6e] external/envoy/source/common/http/conn_manager_impl.cc:?
 #29: Envoy::Http::ConnectionManagerImpl::doEndStream() [0x56042c47bc38]->[0x2aedc38] external/envoy/source/common/http/conn_manager_impl.cc:184
 (inlined by) external/envoy/source/common/http/conn_manager_impl.cc:220
 #30: Envoy::Http::ConnectionManagerImpl::ActiveStream::encodeDataInternal() [0x56042c485e25]->[0x2af7e25] external/envoy/source/common/http/conn_manager_impl.cc:1862
 #31: Envoy::Http::ConnectionManagerImpl::ActiveStream::encodeData() [0x56042c485c57]->[0x2af7c57] external/envoy/source/common/http/conn_manager_impl.cc:1843
 #32: Envoy::Router::UpstreamRequest::decodeData() [0x56042c420bbe]->[0x2a92bbe] bazel-out/k8-opt/bin/external/envoy/source/common/common/_virtual_includes/scope_tracker/common/common/scope_tracker.h:20
 (inlined by) external/envoy/source/common/router/upstream_request.cc:134
 #33: Envoy::Http::ResponseDecoderWrapper::decodeData() [0x56042c38d3db]->[0x29ff3db] bazel-out/k8-opt/bin/external/envoy/source/common/http/_virtual_includes/codec_wrappers_lib/common/http/codec_wrappers.h:37
 #34: Envoy::Http::ResponseDecoderWrapper::decodeData() [0x56042c38d3db]->[0x29ff3db] bazel-out/k8-opt/bin/external/envoy/source/common/http/_virtual_includes/codec_wrappers_lib/common/http/codec_wrappers.h:37
 #35: Envoy::Http::Http1::ClientConnectionImpl::onMessageComplete() [0x56042c4961b1]->[0x2b081b1] bazel-out/k8-opt/bin/external/envoy/source/common/buffer/_virtual_includes/buffer_lib/common/buffer/buffer_impl.h:523
 (inlined by) external/envoy/source/common/http/http1/codec_impl.cc:1079
 #36: Envoy::Http::Http1::ConnectionImpl::onMessageCompleteBase() [0x56042c4937bb]->[0x2b057bb] external/envoy/source/common/http/http1/codec_impl.cc:684
 #37: Envoy::Http::Http1::ConnectionImpl::$_8::__invoke() [0x56042c497f2d]->[0x2b09f2d] external/envoy/source/common/http/http1/codec_impl.cc:393
 #38: http_parser_execute [0x56042c6623ac]->[0x2cd43ac] ??:?
 #39: Envoy::Http::Http1::ConnectionImpl::dispatchSlice() [0x56042c492aea]->[0x2b04aea] external/envoy/source/common/http/http1/codec_impl.cc:527
 #40: Envoy::Http::Http1::ConnectionImpl::dispatch() [0x56042c49285f]->[0x2b0485f] external/envoy/source/common/http/http1/codec_impl.cc:503
 #41: Envoy::Http::CodecClient::onData() [0x56042c3fd048]->[0x2a6f048] external/envoy/source/common/http/codec_client.cc:143
 #42: Envoy::Http::CodecClient::CodecReadFilter::onData() [0x56042c3fdd9d]->[0x2a6fd9d] bazel-out/k8-opt/bin/external/envoy/source/common/http/_virtual_includes/codec_client_lib/common/http/codec_client.h:177
 #43: Envoy::Network::FilterManagerImpl::onContinueReading() [0x56042c21cc13]->[0x288ec13] external/envoy/source/common/network/filter_manager_impl.cc:67
 #44: Envoy::Network::ConnectionImpl::onReadReady() [0x56042c218d55]->[0x288ad55] external/envoy/source/common/network/connection_impl.cc:554
 #45: Envoy::Network::ConnectionImpl::onFileEvent() [0x56042c2180ad]->[0x288a0ad] external/envoy/source/common/network/connection_impl.cc:526
 #46: Envoy::Event::FileEventImpl::assignEvents()::$_0::__invoke() [0x56042c2130d6]->[0x28850d6] external/envoy/source/common/event/file_event_impl.cc:50
 #47: event_process_active_single_queue [0x56042c65bfcb]->[0x2ccdfcb] event.c:?
 #48: event_base_loop [0x56042c65a85e]->[0x2ccc85e] ??:?
 #49: Envoy::Server::WorkerImpl::threadRoutine() [0x56042c206f14]->[0x2878f14] bazel-out/k8-opt/bin/external/envoy/source/common/common/_virtual_includes/minimal_logger_lib/common/common/logger.h:278
 (inlined by) external/envoy/source/server/worker_impl.cc:122
```

And an Istio 1.9.0 crash:
```
Caught Segmentation fault, suspect faulting address 0x60
Backtrace (use tools/stack_decode.py to get line numbers):
Envoy version: 298ff36b2d43794816f7d8cdc5461bf6eed71bba/1.17.0/Clean/RELEASE/BoringSSL
#0: __restore_rt [0x7f60fe64d980]
#1: Envoy::Http::AsyncRequestImpl::onComplete() [0x5577653b6879]
#2: Envoy::Http::AsyncStreamImpl::encodeData() [0x5577653b6170]
#3: Envoy::Http::Utility::sendLocalReply() [0x55776549f748]
#4: Envoy::Http::AsyncStreamImpl::sendLocalReply() [0x5577653b7db2]
#5: Envoy::Router::Filter::sendNoHealthyUpstreamResponse() [0x5577653dac6e]
#6: Envoy::Router::Filter::decodeHeaders() [0x5577653d8ac1]
#7: Envoy::Http::AsyncStreamImpl::sendHeaders() [0x5577653b65c6]
#8: Envoy::Http::AsyncRequestImpl::initialize() [0x5577653b51ef]
#9: Envoy::Http::AsyncClientImpl::send() [0x5577653b4c4b]
#10: Envoy::Extensions::Common::Wasm::Context::httpCall() [0x5577640e7221]
#11: proxy_wasm::exports::http_call() [0x5577644653bb]
#12: proxy_wasm::ConvertFunctionWordToUint32<>::convertFunctionWordToUint32() [0x55776446f7f6]
#13: proxy_wasm::(anonymous namespace)::V8::registerHostFunctionImpl<>()::{lambda()#1}::__invoke() [0x55776445c8ef]
#14: wasm::FuncData::v8_callback() [0x55776447dd2f]
#15: [0xaaad82d64d6]
#16: [0xaaad82b7ce5]
#17: [0xaaad82b78cf]
#18: [0xaaad82b45ce]
#19: [0xaaad82b44d8]
#20: [0xaaad82b43dc]
#21: [0xaaad82acb96]
#22: [0x364000086570]
#23: v8::internal::Execution::CallWasm() [0x557764570939]
#24: wasm::Func::call() [0x55776447d085]
#25: std::__1::__function::__func<>::operator()() [0x5577644541c3]
#26: proxy_wasm::ContextBase::onDone() [0x5577644618f8]
#27: Envoy::Extensions::Common::Wasm::Context::onDestroy() [0x5577640ebdad]
#28: Envoy::Http::ConnectionManagerImpl::doDeferredStreamDestroy() [0x55776533b5ae]
#29: Envoy::Http::ConnectionManagerImpl::ActiveStream::onResetStream() [0x5577653452d7]
#30: Envoy::Http::ConnectionManagerImpl::resetAllStreams() [0x55776533d6b2]
#31: Envoy::Http::ConnectionManagerImpl::doConnectionClose() [0x55776533af24]
#32: Envoy::Network::ConnectionImplBase::raiseConnectionEvent() [0x557765046c7b]
#33: Envoy::Network::ConnectionImpl::raiseEvent() [0x55776503daef]
#34: Envoy::Network::ConnectionImpl::closeSocket() [0x55776503d78b]
#35: Envoy::Network::ConnectionImpl::onFileEvent() [0x55776504026c]
#36: std::__1::__function::__func<>::operator()() [0x5577650348c1]
#37: Envoy::Event::FileEventImpl::assignEvents()::$_1::__invoke() [0x5577650356cc]
#38: event_process_active_single_queue [0x5577654c5d0b]
#39: event_base_loop [0x5577654c466e]
#40: Envoy::Server::WorkerImpl::threadRoutine() [0x557765025473]
#41: Envoy::Thread::ThreadImplPosix::ThreadImplPosix()::{lambda()#1}::__invoke() [0x5577656ab663]
#42: start_thread [0x7f60fe6426db]
2021-03-03T05:07:13.223457Z	error	Epoch 0 exited with error: signal: segmentation fault
2021-03-03T05:07:13.223476Z	info	No more active epochs, terminating
```

**Version** (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version --short` if you used Helm)

```
client version: 1.9.0
control plane version: 1.9.0
data plane version: 1.9.0

Client Version: v1.19.3
Server Version: v1.18.9-eks-d1db3c
```

**How was Istio installed?**

istioctl

**Environment where the bug was observed (cloud vendor, OS, etc)**

AWS EKS k8s 1.18

Additionally, please consider running `istioctl bug-report` and attach the generated cluster-state tarball to this issue.
Refer [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) for more details.

**Comment by bianpengyuan** [comment]
Is there any chance that you could provide a minimum code snippet to reproduce this?

**Comment by chandlernine** [comment]
A sketch of how to reproduce this is https://github.com/chandlernine/proxy-runtime/pull/1/files

**Comment by bianpengyuan** [comment]
You returned true in ondone: https://github.com/chandlernine/proxy-runtime/pull/1/files#diff-50ad1f311e4dad222656d7e89383a67edb9ee3ff1a531450efd73ce44545f964R53, which indicates that the Auth context is safe to be destroyed. You should return false instead, and call `proxy_done` inside http finishing callbacks.

**Comment by bianpengyuan** [comment]
You can find a similar example here, where a grpc call is made in ondone https://github.com/istio-ecosystem/wasm-extensions/blob/32096c8f6bb8f3c3d0702f997910186b68d9e649/extensions/grpc_logging/plugin.cc#L86 and then proxy done is called in grpc callbacks: https://github.com/istio-ecosystem/wasm-extensions/blob/32096c8f6bb8f3c3d0702f997910186b68d9e649/extensions/grpc_logging/plugin.cc#L63

**Comment by chandlernine** [comment]
I am returning true in onDone of a non-root context because my http call is made on a root context. Why does my non-root context need to persist?

**Comment by bianpengyuan** [comment]
I dont know much about rust API, but I saw that this is passed in to the httpCall, so are you sure the stream context is not used in any way?

**Comment by chandlernine** [comment]
In assemblyscript, that context (callback.origin_context) is indeed used later in ways that would be disastrous if the non-root context had been reclaimed:
```
  onHttpCallResponse(token: u32, headers: u32, body_size: u32, trailers: u32): void {
      log(LogLevelValues.debug, "onHttpCallResponse: calling callback for context id: " + callback.origin_context.context_id.toString());
      callback.cb(callback.origin_context, headers, body_size, trailers);
```
My crashes do not seem to call proxy_on_http_call_response, though. Before that can happen, some pointer related to the root context or an HTTP call from it is invalid.

But I bet you've solved my problem, so let me go try using proxy_done and returning false from onDone(). I'll reopen if I can still reproduce the crash.

**Comment by chandlernine** [comment]
I changed proxy_on_done to return false instead of true, but on both Istio 1.6.13 and 1.9.0 it resulted in exactly the same backtraces. The C++ code ignores the return value from a non-root context; some documentation states that the WASM must return 'Done' for a non-root context, which matches that code.

So the only interesting 'return false' is for a root context, and I think it had no effect because the VM is not shutting down and therefore proxy_on_done is not called on any root context.

**Comment by bianpengyuan** [comment]
Could you provide c++ code? It will be more understandable for me :)

**Comment by chandlernine** [comment]
For Istio 1.6.13:
```
// For stream Contexts, onDone always returns true.
bool onDoneBase() override {
  onDone();
  return true;
}
....
/**
 * For stream contexts, called when the stream has completed. Note: if applicable proxy_on_log() is
 * called after proxy_on_done() and before proxy_on_delete(). For root contexts, proxy_on_done() is
 * called when the VM is going to shutdown.
 * @param context_id is an identifier the context.
 * @return non-zero to indicate that this context is done. Stream contexts must return non-zero.
 * Root contexts may return zero to defer the VM shutdown and the proxy_on_delete call until after a
 * future proxy_done() call by the root context.
 */
enum class WasmOnDoneResult : uint32_t {
  Done = 0,
  NotDone = 1,
}
extern "C" WasmOnDoneResult proxy_on_done(uint32_t context_id);
```

**Comment by bianpengyuan** [comment]
I mean could you provide your C++ code that causes crash? Or you are saying you are just reading the C++ sdk code and not writing your plugin with C++?

If ondone aways return true, which means it won't handle asynchronous calls, you probably should not pass in `this` in the http call. I guess that is used as context to execute http callbacks. You probably want to pass in this.root_context instead. Again I know nothing about asmjs sdk, so take my words with grain of salt. And you will probably get better support by opening an issue at solo's repo.

**Comment by PiotrSikora** [comment]
This is fixed in https://github.com/envoyproxy/envoy/pull/15311 (pending merge upstream).

**Comment by chandlernine** [comment]
Thank you both!

Given that the fix also solves https://github.com/envoyproxy/envoy/issues/14878, I'm curious if I could work around this issue by switching the internal cluster from EDS to something else?

(For the record, the code I provided is istio-proxy's Envoy, not a WASM SDK. In 1.9.0 it would be proxy-wasm-cpp-host, in part; in 1.6.13 it was all in github.com/istio/envoy.)

**Comment by PiotrSikora** [comment]
> Given that the fix also solves [envoyproxy/envoy#14878](https://github.com/envoyproxy/envoy/issues/14878), I'm curious if I could work around this issue by switching the internal cluster from EDS to something else?

I don't believe you could.
